README.md
# Download link for v1.03 (latest)
If updating from v0.85 or lower, delete all previous files apart from your config. Run the EXE in the same folder as your config.
The directoy should look as follows:
* GW2TT.exe
* exampleConfig.ini
* config.ini

If you do not have a config.ini, just run the application once to generate it.
* https://github.com/ndj710/GW2-Trading-post-tracker/raw/main/downloads/GW2TT%20v1.03.zip

# Note:
This is an ongoing project, I am making an application for windows on behalf of a friend. I am experimenting with tkinter and other libraries.

# GW2TT
GWT22 is a windows application which allows the user to enter an item ID and then set a price for that item, the application will make periodic API calls to fetch the item price, if the price the user set is met, the user will be emailed a notification.

## Screenshots
<div>
  <h5>Main page</h5>
  <img src="/screenshots/Mainpage.png?raw=true" width="639" height="373"/>
</div>


### Changelog:
#### Version 1.03

Changes:
*	Can now drag items and swap with other items using the hamburger on left (Currently does NOT save positions on app close)

Bug fixes:
*	Pressing add on blank item ID entry field will no longer trigger an error in logs
*	Bug where changing price to 0 while tracking item (green bell) would cause a crash


#### Version 1.02

Changes:
*	Adjusted the padding at bottom of notification button to center the textimage
*	Added a state for notifcation button being triggered
*	Adjusted delete button height to match notification button

Bug fixes:
*	Logs showing same error multiple times


#### Version 1.01

Changes:
*	Changed size of setting input fields to match rest of UI
*	Can now list an item twice (one for selling and one for buying)
*	Title now has capital T's

Bug fixes:
*	Removed leading whitespace from item names


#### Version 1.00
Finished all the basic functionality, application is ready for use.
Changes:
*	Added scrollbar (have to use mouse and drag, scrollwheel not currently working)
*	Added link to item data on https://www.gw2bltc.com/ by clicking itemID
*	Removed priceData base, price will just be checked after each API call. No need for hisotoric data

Bug fixes:
*	Removed items that are not listed in the tradepost from being loaded into memory
* 	Issue where deleting one item and adding another would sometimes overlap current items in the watchlist


#### Version 0.85

Changes:
*	Changed item data population method, Item Data no longer needs to be saved in database. Only price data is saved to database


#### Version 0.84

Changes:
*	Cleaned up source code

Bug fixes:
* 	fix issue if opening without images folder
*	Issue with notification bell being red when should be green (sometimes)


#### Version 0.83

Changes:
* 	Changed format of current price
* 	Reduced window size by changing text font
* 	Added placeholder price for current price until first update is received
* 	Can no longer set item to be tracked if no price has been entered
* 	Added some checks and blocks to stop tracking of items with target price of 0

Bug fixes:
* 	Fixed a graphical glitch if no items are in the watchlist on boot up
* 	Fixed bug where 'Item ID' is not reporting incorrect ID
* 	Fixed placeholder for target price entry boxes not showing correctly

#### Version 0.82

Changes:
* 	Price now in colours and displays images for gold, silver, copper

Bug fixes:
* 	Fixed the ungraceful closing of update thread
* 	Fixed incorrect price conversion (+- 1 copper)


changelog.txt
Version 0.83 Changelog

Changes:
	-Changed format of current price
	-Reduced window size by changing text font
	-Added placeholder price for current price until first update is received
	-Can no longer set item to be tracked if no price has been entered
	-Added some checks and blocks to stop tracking of items with target price of 0

Buf fixes:
	-Fixed a graphical glitch if no items are in the watchlist on boot up
	-Fixed bug where 'Item ID' is not reporting incorrect ID
	-Fixed placeholder for target price entry boxes not showing correctly



Version 0.82 Changelog

Changes:
	-Price now in colours and displays images for gold, silver, copper

Bug fixes:
	-Fixed the ungraceful closing of update thread
	-Fixed incorrect price conversion (+- 1 copper)



src/tradePostTracker.py
import requests
import time
import threading
import configparser
import traceback
import email
import smtplib
from stoppableThread import StoppableThread
import logging

logging.getLogger("urllib3").setLevel(logging.WARNING)

class TradePostTracker:
    def __init__(self):
        self.loading = True
        self.threads = []
        self.error = ''
        self.timer = 10
        self.itemIds = []
        self.itemData = {}
        self.server = ''
        self.port = 587
        self.sender = ''
        self.password = ''
        self.receiver = ''
        self.currentPrices = {}
        self.speaker = []
        self.alertButtons = []
    
    def addToConfig(self):
        try:
            config = configparser.ConfigParser()
            config.read('config.ini')
            strToAdd = ''
            for item in self.itemIds:
                if item != None:
                    strToAdd += '{}|{}|{}|{}'.format(item[0], item[1], item[2], item[3]) + ', '
            config.set('Configuration','ITEMS', strToAdd.strip(', '))
            config.set('Email','server', self.server)
            config.set('Email','port', str(self.port))
            config.set('Email','sender', self.sender)
            config.set('Email','password', self.password)
            config.set('Email','receiver', self.receiver)
            with open('config.ini', 'w') as newini:
                config.write(newini)
        except Exception as e:
            self.error = traceback.format_exc()
    
    def parseConfig(self):
        try:
            config = configparser.ConfigParser()
            config.read('config.ini')
            if config.sections() == []:
                config['Configuration'] = {'TIMER': '10', 'ITEMS': '9482|buy|100|f, 24|sell|10|t'}
                config['Email'] = {'server': 'smtp.office365.com', 'port': 587, 'sender': 'examplesender@hotmail.com', 'password': 'examplepassword', 'receiver': 'examplereceiver@emaildomain.com'}
                with open('config.ini', 'w') as configfile: config.write(configfile)
            self.timer = int(config['Configuration']['TIMER'])
            confItems = config['Configuration']['ITEMS'].replace(' ', '').split(',')
            self.server = config['Email']['server']
            self.port = int(config['Email']['port'])
            self.sender = config['Email']['sender']
            self.password = config['Email']['password']
            self.receiver = config['Email']['receiver']
            for i in confItems:
                i = i.split('|')
                if i != ['']:
                    errorFlag = False
                    for x in self.itemIds:
                        if x[0] == i[0] and x[1] == i[1]:
                            errorFlag = True
                    if errorFlag == True:
                        self.error = 'Too many duplicates in config file, see default for examples'     
                        print(self.error)
                        return
                    self.itemIds.append(i)
        except Exception as e:
            if e == 'Configuration':
                self.error = 'Error in config file, see default for examples'      
            elif e == 'Email':
                self.error = 'Error in config file, see default for examples'
            else:
                self.error = traceback.format_exc()
            
    def loadItemData(self):
        try:
            url = 'https://api.datawars2.ie/gw2/v1/items/json?filter=$fields=id,name,buy_price,sell_price'
            req = requests.get(url).json()
            for fullitem in req:
                if 'id' in fullitem and 'name' in fullitem and 'buy_price' in fullitem and 'sell_price' in fullitem:
                    self.itemData[fullitem['id']] = fullitem['name']
        except Exception as e:
            self.error = traceback.format_exc()

    def sendEmail(self, item):
        try:
            self.speaker[item[4]].set(u"\U0001F4E9")
            self.alertButtons[item[4]].configure(fg_color=("#F28C28"))
            msg = email.message_from_string('Item {} ({}) has hit the alert threshold'.format(self.itemData[int(item[0])], str(item[0])))
            msg['From'] = self.sender
            msg['To'] = self.receiver
            msg['Subject'] = "GW2 {} ({}) PA".format(self.itemData[int(item[0])], str(item[0]))
            s = smtplib.SMTP(self.server,self.port)
            s.ehlo()
            s.starttls()
            s.ehlo()
            s.login(self.sender, self.password)
            s.sendmail(self.sender, self.receiver, msg.as_string())
            s.quit()         
        except Exception as e:
            self.error = traceback.format_exc()
    
    
    def convertPrice(self, price):
        try:
            gold = '0'
            silver = '0'
            copper = '0'
            if len(price) >= 5:
                gold = price[:-4]
            if len(price) >= 3:
                silver = price[-4:-2]
                if silver[0] == '0':
                    silver = silver[1]
            if len(price) >= 2:
                copper = price[-2:]
                if copper[0] == '0':
                    copper = copper[1]
            return ('{}g {}s {}c'.format(gold, silver, copper), (gold, silver, copper))
        except Exception as e:
            self.error = traceback.format_exc()
    
    def getCurrentPrice(self, itemId, df):
        try:
            buy = str(df['BuyPrice'])
            sell = str(df['SellPrice'])
            buyPrice = self.convertPrice(buy)
            sellPrice = self.convertPrice(sell)
            self.currentPrices[str(itemId)][0][0].set(buyPrice[1][0])
            self.currentPrices[str(itemId)][0][1].set(buyPrice[1][1])
            self.currentPrices[str(itemId)][0][2].set(buyPrice[1][2])
            self.currentPrices[str(itemId)][1][0].set(sellPrice[1][0])
            self.currentPrices[str(itemId)][1][1].set(sellPrice[1][1])
            self.currentPrices[str(itemId)][1][2].set(sellPrice[1][2])
        except Exception as e:
            self.error = traceback.format_exc()
            
    def startUpdate(self):
        self.threads.append(StoppableThread(target=self.getItemPrices))
        self.threads[0].start()
    
    def getItemPrices(self):
        while True:    
            try:
                print('running')
                if self.error != '':
                    logging.basicConfig(filename='./logs/updateThreadCrash.log', encoding='utf-8', level=logging.DEBUG)
                    logging.error(self.error)   
                    self.error = ''
                count = 0
                while count < self.timer:
                    if self.threads[0].stopped() == True:
                        print('Thread closed')
                        return                    
                    time.sleep(1)
                    count += 1
                if self.loading == False:
                    url = 'https://api.guildwars2.com/v2/commerce/prices?ids='
                    for item in self.itemIds:
                        if item != None:
                            url += str(item[0]) + ','
                    self.create_frame(requests.get(url).json())
            except Exception as e:
                self.error = traceback.format_exc()

    def create_frame(self, data):
        try:
            for item in data:
                if item == 'text':
                    print('passing')
                    continue
                df = {'id': item['id'], 'BuyPrice': item['buys']['unit_price'], 'SellPrice': item['sells']['unit_price']}
                itemId = df['id']
                for i in self.itemIds:
                    if i != None and i[0] == str(itemId):
                        targetPrice =  int(i[2])
                        if self.speaker[i[4]].get() == u"\U0001F514" and targetPrice != 0:
                            if i[1] == 'buy':
                                currentPrice = int(df['BuyPrice'])
                                if currentPrice <= targetPrice:
                                    self.sendEmail(i)
                            elif i[1] == 'sell':
                                currentPrice = int(df['SellPrice'])
                                if currentPrice >= targetPrice:
                                    self.sendEmail(i)    
                self.getCurrentPrice(itemId, df)
        except Exception as e:
            self.error = traceback.format_exc()

TradePostTracker.py
import requests
import sqlalchemy
import time
import pandas as pd
pd.options.mode.chained_assignment = None
from datetime import datetime
import threading
import configparser
import traceback
import email
import smtplib
from stoppableThread import StoppableThread


class TradePostTracker:
    def __init__(self):
        self.loading = True
        self.threads = []
        self.error = ''
        self.timer = 10
        self.itemIds = []
        self.itemData = {}
        self.server = ''
        self.port = 587
        self.sender = ''
        self.password = ''
        self.receiver = ''
        self.engine = sqlalchemy.create_engine('sqlite:///itemData.db')
        self.currentPrices = {}
        self.speaker = {}
        self.alertButtons = {}
    
    def addToConfig(self):
        try:
            config = configparser.ConfigParser()
            config.read('config.ini')
            strToAdd = ''
            for item in self.itemIds:
                if item != None:
                    strToAdd += '{}|{}|{}|{}'.format(item[0], item[1], item[2], item[3]) + ', '
            config.set('Configuration','ITEMS', strToAdd.strip(', '))
            config.set('Email','server', self.server)
            config.set('Email','port', str(self.port))
            config.set('Email','sender', self.sender)
            config.set('Email','password', self.password)
            config.set('Email','receiver', self.receiver)
            with open('config.ini', 'w') as newini:
                config.write(newini)
        except Exception as e:
            self.error = 'Error in TPTaddToConfig'
            print(self.error + ' ', e)
    
    def parseConfig(self):
        try:
            config = configparser.ConfigParser()
            config.read('config.ini')
            if config.sections() == []:
                config['Configuration'] = {'TIMER': '10', 'ITEMS': '9482|buy|100|f, 24|sell|10|t'}
                config['Email'] = {'server': 'smtp.office365.com', 'port': 587, 'sender': 'examplesender@hotmail.com', 'password': 'examplepassword', 'receiver': 'examplereceiver@emaildomain.com'}
                with open('config.ini', 'w') as configfile: config.write(configfile)
            self.timer = int(config['Configuration']['TIMER'])
            confItems = config['Configuration']['ITEMS'].replace(' ', '').split(',')
            self.server = config['Email']['server']
            self.port = int(config['Email']['port'])
            self.sender = config['Email']['sender']
            self.password = config['Email']['password']
            self.receiver = config['Email']['receiver']
            for i in confItems:
                i = i.split('|')
                if i != ['']:
                    self.itemIds.append(i)
        except Exception as e:
            if e == 'Configuration':
                self.error = 'Error in config file, see default for examples'      
            elif e == 'Email':
                self.error = 'Error in config file, see default for examples'
            else:
                self.error = 'Error in TPTparseConfig'
            print(self.error + ' ', e)
    
    def loadItemData(self):
        try:
            def load(itemReq):
                url = 'https://api.guildwars2.com/v2/items?ids='
                nameId = {'id': [], 'name': []}
                for index, item in enumerate(itemReq):
                    if index % 100 != 0:
                        url += str(item) + ','
                    else:
                        url += str(item)
                        req = requests.get(url).json()
                        if "text" not in req:
                            for fullitem in req:
                                nameId['id'].append(fullitem['id'])
                                nameId["name"].append(fullitem['name'])
                            df = pd.DataFrame.from_dict(nameId)
                            df.to_sql('Items', self.engine, if_exists='append', index=False)
                            url = 'https://api.guildwars2.com/v2/items?ids='
                            nameId = {'id': [], 'name': []}
                req = requests.get(url).json()
                if "text" not in req:
                    for fullitem in req:
                        nameId['id'].append(fullitem['id'])
                        nameId["name"].append(fullitem['name'])
                    url = 'https://api.guildwars2.com/v2/items?ids='        
                    df = pd.DataFrame.from_dict(nameId)
                    df.to_sql('Items', self.engine, if_exists='append', index=False)
            connection = self.engine.connect()
            metadata = sqlalchemy.MetaData()
            tab = sqlalchemy.Table('Items', metadata, autoload=True, autoload_with=self.engine)
            query = sqlalchemy.select([tab])
            result = connection.execute(query)
            result = result.fetchall()
            for i in result:
                self.itemData[i[0]] = i[1]
            itemReq = requests.get('https://api.guildwars2.com/v2/commerce/prices').json()
            notAdded = []
            for i in itemReq:
                try:
                    self.itemData[i]
                except:
                    notAdded.append(i)
            load(notAdded)
        except Exception as e:
            if str(e) == 'Items':
                itemReq = requests.get('https://api.guildwars2.com/v2/commerce/prices').json()
                load(itemReq)
                connection = self.engine.connect()
                metadata = sqlalchemy.MetaData()
                tab = sqlalchemy.Table('Items', metadata, autoload=True, autoload_with=self.engine)
                query = sqlalchemy.select([tab])
                result = connection.execute(query)
                result = result.fetchall()
                for i in result:
                    self.itemData[i[0]] = i[1]
            else:
                self.error = 'Error in TPTloadItemData'
                print(self.error + ' ', e)
                print(traceback.format_exc())
        
    def sendEmail(self, item):
        try:
            self.speaker[item[0]].set(u"\U0001F515")
            self.alertButtons[item[0]].configure(fg_color=("#A7171A"))               
            msg = email.message_from_string('Item {} ({}) has hit the alert threshold'.format(self.itemData[int(item[0])], str(item[0])))
            msg['From'] = self.sender
            msg['To'] = self.receiver
            msg['Subject'] = "GW2 {} ({}) PA".format(self.itemData[int(item[0])], str(item[0]))
            s = smtplib.SMTP(self.server,self.port)
            s.ehlo()
            s.starttls()
            s.ehlo()
            s.login(self.sender, self.password)
            s.sendmail(self.sender, self.receiver, msg.as_string())
            s.quit()         
        except Exception as e:
            self.error = 'Error in TPTsendEmail'
            print(self.error + ' ', e)
    
    
    def convertPrice(self, price):
        gold = '0'
        silver = '0'
        copper = '0'
        if len(price) >= 5:
            gold = price[:-4]
        if len(price) >= 3:
            silver = price[-4:-2]
            if silver[0] == '0':
                silver = silver[1]
        if len(price) >= 2:
            copper = price[-2:]
            if copper[0] == '0':
                copper = copper[1]
        
        return ('{}g {}s {}c'.format(gold, silver, copper), (gold, silver, copper))  
    
    def getCurrentPrice(self, itemId, table_df):
        try:
            buy = str(table_df['BuyPrice'][len(table_df)-1])
            sell = str(table_df['SellPrice'][len(table_df)-1])
            buyPrice = self.convertPrice(buy)
            sellPrice = self.convertPrice(sell)
            self.currentPrices[str(itemId)][0][0].set(buyPrice[1][0])
            self.currentPrices[str(itemId)][0][1].set(buyPrice[1][1])
            self.currentPrices[str(itemId)][0][2].set(buyPrice[1][2])
            self.currentPrices[str(itemId)][1][0].set(sellPrice[1][0])
            self.currentPrices[str(itemId)][1][1].set(sellPrice[1][1])
            self.currentPrices[str(itemId)][1][2].set(sellPrice[1][2])
        except Exception as e:
            self.error = 'Error in TPTgetCurrentPrice'
            print(self.error + ' ', e)
            
    def startUpdate(self):
        self.threads.append(StoppableThread(target=self.getItemPrices))
        self.threads[0].start()
    
    def getItemPrices(self):
        while True:
            if self.threads[0].stopped() == True:
                print('Thread closed')
                break        
            try:
                print('running')
                time.sleep(self.timer)
                if self.loading == False:
                    url = 'https://api.guildwars2.com/v2/commerce/prices?ids='
                    for item in self.itemIds:
                        if item != None:
                            url += str(item[0]) + ','
                    self.create_frame(requests.get(url).json())
            except Exception as e:
                self.error = 'Error in TPTgetItemPrice'
                print(self.error + ' ', e)

    def create_frame(self, data):
        try:
            for item in data:
                if item == 'text':
                    print('passing')
                    continue
                df = pd.DataFrame([item])
                date = datetime.now()
                df['Time'] = date.strftime("%d-%m-%Y, %H:%M:%S")
                df['BuyPrice'] = df['buys'][0]['unit_price']
                df['BuyAmount'] = df['buys'][0]['quantity']
                df['SellPrice'] = df['sells'][0]['unit_price']
                df['SellAmount'] = df['sells'][0]['quantity']
                df = df.loc[:, ['id', 'whitelisted', 'BuyPrice', 'BuyAmount', 'SellPrice', 'SellAmount', 'Time']]
                itemId = df['id'][0]
                
                for i in self.itemIds:
                    if i != None and i[0] == str(itemId):
                        targetPrice =  int(i[2])
                        print('targetPrice: ', targetPrice)
                        print(self.speaker[str(itemId)].get())
                        if self.speaker[str(itemId)].get() != u"\U0001F515" and targetPrice != 0:
                            try:
                                table_df = pd.read_sql_table(str(itemId), self.engine)
                                if i[1] == 'buy':
                                    currentPrice = int(df['BuyPrice'][0])
                                    if currentPrice <= targetPrice:
                                        self.sendEmail(i)
                                elif i[1] == 'sell':
                                    currentPrice = int(df['SellPrice'][0])
                                    if currentPrice >= targetPrice:
                                        self.sendEmail(i)
                            except Exception as e:
                                self.error = 'Error in TPTcreateframe inner loop'
                                print(self.error + ' ', e)
                                pass
                        break            
                
                df.to_sql(str(itemId), self.engine, if_exists='append', index=False)
                self.getCurrentPrice(itemId, df)
        except Exception as e:
            pass